rfcs/004 - Continuous Two-Way Synchronization RFC.md

```diff
@@ -0,0 +1,160 @@
+
+SELECT iasql_drop_enforce('table_name', 5); -- Drop the enforcement of a particular row. This one, at least, is unambiguous.
+
+SELECT iasql_drop_enforce_all(); -- Restores normal behavior.
+
+SELECT iasql_track('table_name'); -- Makes the table effectively read-only.
+
+SELECT iasql_track('table_name', 5); -- Makes a particular row read-only.
+
+SELECT iasql_track_all(); -- Makes the DB read-only
+
+SELECT iasql_drop_track('table_name'); -- Drops tracking for the table. Similar issue to above: should this affect row-level tracking in the same table?
+
+SELECT iasql_drop_track('table_name', 5); -- Drops tracking for a particular row.
+
+SELECT iasql_drop_track_all(); -- Restores normal behavior
+```
+
+While entering the transaction mode introduces a synchronous point (such that `CALL iasql_commit();` blocks until the apply/sync operation is complete), there may be times where a user made changes in "normal" two-way mode but they now need to wait for resources to actually be created. A proposed helper function would trigger a blocking call until a user condition is met. One possible syntax for this is:
+
+```sql
+SELECT iasql_wait_for('table_name', 5, 'cloud_id IS NOT NULL'); -- Default version
+SELECT iasql_wait_for('table_name', 5, 'cloud_id IS NOT NULL', 30); -- Optional argument for maximum # of seconds to wait before erroring (defaulting to 5 minutes or something)
+```
+
+Similar to the enforce and track functions, it is essentially executing (on another postgres thread/process) a particular select statement in a loop with a sleep thrown in until success or it errors out if the condition is never met. That select statement would be something like:
+
+```sql
+SELECT count(*) > 0 FROM `table_name` WHERE id = `5` AND `cloud_id IS NOT NULL`
+```
+
+The parts in backticks are from the user-provided values. This is SQL injection, but it's their own database and we already allow arbitrary SQL statements to run on the database, so it doesn't affect the security situation in the slightest, even if it does immediately cause heart palpitations while looking at it. ;)
+
+The final recommendation is to keep `iasql_apply` and `iasql_sync` as functions that can be called, but make them no-ops that warn that they don't do anything anymore, and that at some time in the future these functions will be dropped. The reasoning behind this is that this transition will take some time and there may be IaSQL snippets out there that do continue working after this transition assuming they didn't call either of those functions. We can track the number of actual invocations in these functions by having the engine-side implementation simply fire off an alert to us. If we see little-to-no triggering by end users, we could drop them sooner rather than later.
```

Comment:
we do need to drop `iasql_apply_preview` and `iasql_sync_preview` ahead of the v0.0.1 release since they won't be backwards compatible, right?

Reply:
If we release v0.1 without the two-way mode RFC work implemented, we *probably* still want them around for tutorials to use them, right?

Then it would make sense to have these functions just `raise exception` that they no longer work and that they should read the docs on IaSQL transactions.